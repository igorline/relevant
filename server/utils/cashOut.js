import Earnings from '../api/earnings/earnings.model';

export async function logCashOut(user, cashOutAmt, next) {
  try {
    const earnings = await Earnings.find({
      user: user._id,
      status: 'paidout',
      earned: { $gt: 0 }
    });
    const totalPreviousPaidout = earnings.reduce(getTotalEarnings, 0);
    const cashOutLog = new Earnings({
      user: user._id,
      cashOutAttempt: true,
      cashOutAmt,
      prevBalance: user.balance,
      endBalance: user.balance - cashOutAmt,
      totalPreviousPaidout,
      legacyAirdrop: user.legacyAirdrop,
      referralTokens: user.referralTokens,
      airdropTokens: user.airdropTokens,
      status: 'in progress',
      source: 'wallet'
    });
    await cashOutLog.save();
    return cashOutLog;
  } catch (err) {
    return next(new Error('Error logging cashOut', err));
  }
}

export function getTotalEarnings(total, currentVal) {
  total += currentVal.earned;
  return total;
}
