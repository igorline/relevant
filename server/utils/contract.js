import { Contract, ethers } from 'ethers';

const erc20abi = [
  {
    constant: true,
    inputs: [
      {
        name: '_owner',
        type: 'address'
      }
    ],
    name: 'balanceOf',
    outputs: [
      {
        name: 'balance',
        type: 'uint256'
      }
    ],
    payable: false,
    type: 'function'
  }
];

export const getTokenBalance = async ({ address, tokenAddress }) => {
  const provider = ethers.getDefaultProvider('mainnet');
  const contract = new Contract(tokenAddress, erc20abi, provider);
  const balance = await contract.balanceOf(address);
  return balance.toString();
};
