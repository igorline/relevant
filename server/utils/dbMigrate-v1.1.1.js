import CommunityMember from 'server/api/community/community.member.model';
import Community from 'server/api/community/community.model';

async function markDeletedMembers() {
  const deletedCom = await Community.find({ inactive: true }, '_id');
  const ids = deletedCom.map(c => c._id);
  return CommunityMember.updateMany(
    { communityId: { $in: ids } },
    { deletedCommunity: true },
    { new: true }
  );
}

async function updateDB() {
  try {
    await markDeletedMembers();
  } catch (err) {
    console.log(err); // eslint-disable-line
  }
}

updateDB();
