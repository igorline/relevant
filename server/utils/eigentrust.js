
function eigenTrust(users, params) {
  const defaults = {
    alpha: 0.95,
    error: 0.0001,
    maxIter: 200,
    initialTrust: [],
  };
  let { alpha, error, maxIter, initalTrust } = { ...defaults, ...params };



}

export default eigenTrust;