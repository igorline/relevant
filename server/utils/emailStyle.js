export const notificationStyle = `
<style>
  .post {
    border: 1px solid lightgrey;
    padding: 4px 16px 4px 16px;
    max-width: 600px;
  }
  .post a {
    text-decoration: none;
  }
  .post .head {
    padding-top: 12px;
  }
  .post .body {
    display: block;
    padding: 12px 0;
    text-decoration: none;
    color: black;
  }
  .footer {
    font-size: 10px;
    color: grey;
  }
</style>`;

export const emailStyle = `
<style>
  p, ul {
    font-size: 14px;
    line-height: 22.5px;
    font-family: Georgia;
    max-width: 667px;
  }
  hr {
    border: .5px solid grey;
    max-width: 664px;
    margin: 30px 0;
  }

  p img {
    max-width: 100%;
    margin: auto;
    text-align: center;
  }
  p img.storeButton {
    margin-top: -10px;
    margin-left: -10px;
  }
  p.buttons {
    margin: 30px 0;
  }
  a {
    color: #3E3EFF;
  }
  a span {
    font-weight: bold;
    font-size: 24px;
    border: 3px solid #3E3EFF;
    padding: 15px 20px;
    display: inline-block;
    margin: 15px 0;
    font-family: Arial;
  }
</style>
`;
