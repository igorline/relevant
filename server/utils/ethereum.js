import request from 'request-promise-any';
import { ethers } from 'ethers';
import { INFURA_NETWORK, NETWORK_NUMBER } from 'app/core/config';

const RelevantToken = require('../../app/contracts/RelevantToken.json');

let decimals;
let instance;
let rpcUrl;
let provider;
let network;
let key;
let tokenAddress;
let web3;
let initialized = false;
let wallet;

const pendingTx = {};

process.on('exit', () => {
  // eslint-disable-next-line
  console.log('Un-executed Pending Transactions: ', pendingTx);
});

export const isInitialized = () => initialized;
export const getWeb3 = () => web3;

// SECURITY - this should never by exposed via any APIs!
export const getInstance = () => instance;

export async function init() {
  try {
    // SECURITY - this env var should never by exposed via any APIs!
    if (process.env.NODE_ENV === 'production') {
      key = process.env.OWNER_KEY;
      if (!key) return false;

      provider = ethers.getDefaultProvider(INFURA_NETWORK);
      tokenAddress = RelevantToken.networks[NETWORK_NUMBER]
        ? RelevantToken.networks[NETWORK_NUMBER].address
        : null;
    } else {
      key = process.env.TEST_KEY;
      if (!key) return false;

      rpcUrl = process.env.TEST_RPC;
      provider = new ethers.providers.JsonRpcProvider(rpcUrl);
      network = await provider.getNetwork();
      tokenAddress = RelevantToken.networks[network.chainId].address;
    }

    wallet = new ethers.Wallet(key, provider);
    instance = new ethers.Contract(tokenAddress, RelevantToken.abi, provider);

    // SECURITY - this should never by exposed via any APIs!
    instance = instance.connect(wallet);

    decimals = await instance.decimals();
    initialized = true;
    return true;
  } catch (err) {
    throw err;
  }
}

export async function getBalance(address) {
  if (!instance) return 0;
  const balance = await instance.balanceOf(address);
  return balance.div((10 ** decimals).toString()).toNumber();
}

export async function getParam(param, opt) {
  if (!initialized) await init();

  let value = await instance[param]();
  if (!opt || !opt.noConvert) value = value.div((10 ** decimals).toString());
  if (!opt || !opt.string) value = parseFloat(value.toString());
  return value;
}

export async function getGasPrice() {
  const gasPrice = await request('https://ethgasstation.info/json/ethgasAPI.json');
  const price = JSON.parse(gasPrice);
  console.log('gas price', price.fast); // eslint-disable-line
  return price.fast;
}

// SECURITY - this function should never by exposed via any APIs!
export async function sendTx({ method, args, cancelPendingTx }) {
  try {
    const gasPrice = await getGasPrice();
    const nonce = await provider.getTransactionCount(wallet.address);
    console.log('current nonce', nonce); // eslint-disable-line
    const optNonce = cancelPendingTx ? { nonce } : {};
    const options = {
      gasPrice: gasPrice * 1e8,
      gasLimit: 6e6,
      ...optNonce
    };
    const tx = await instance[method](...args, options);
    pendingTx[method] = tx;
    const result = await tx.wait();
    delete pendingTx[method];
    console.log('status:', result.status); // eslint-disable-line
    console.log(`gas used by ${method}: ${result.gasUsed}`); // eslint-disable-line
    return result;
  } catch (err) {
    throw err;
  }
}

function toBN(num) {
  return ethers.utils.parseUnits((num / 10 ** decimals).toString(), decimals);
}

export async function mintRewardTokens() {
  if (!instance) await init();
  const lastMint = await instance.roundsSincleLast();
  if (lastMint.toNumber() === 0) return null;
  return sendTx({ method: 'releaseTokens', args: [], overWritePending: true });
}

export async function allocateRewards(_amount) {
  return sendTx({ method: 'allocateRewards', args: [toBN(_amount)] });
}

export async function allocateAirdrops(_amount) {
  return sendTx({ method: 'allocateAirdrops', args: [toBN(_amount)] });
}

export async function getNonce(_account) {
  const nonce = await instance.nonceOf(_account);
  return nonce.toNumber();
}

export async function sign(_account, _amount) {
  const nonce = await getNonce(_account);
  // const amnt = ethers.utils.bigNumberify((_amount * 10 ** 18).toString());
  const amnt = ethers.utils.parseUnits(_amount.toString(), decimals);
  const hash = ethers.utils.solidityKeccak256(
    ['uint256', 'address', 'uint256'],
    [amnt.toString(), _account, nonce]
  );
  const ethSig = await wallet.signMessage(ethers.utils.arrayify(hash));
  return ethSig;
}
