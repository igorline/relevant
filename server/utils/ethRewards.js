import User from '../api/user/user.model';
import Invest from '../api/invest/invest.model';
import apnData from '../pushNotifications';
import Notification from '../api/notification/notification.model';
import Earnings from '../api/earnings/earnings.model';
import Community from '../api/community/community.model';
import * as Eth from './ethereum';
import { SHARE_DECAY, MINIMUM_RANK } from '../config/globalConstants';
import * as numbers from '../../app/utils/numbers';
import PostData from '../api/post/postData.model';
import computePageRank from './pagerankCompute';

/* eslint no-console: 0 */

const queue = require('queue');

const q = queue({ concurrency: 1 });

let totalDistributedRewards;

// ANALYSIS â€” attack scenario community with low-quality posts to bring down the average?
async function computePostPayout(posts, community) {
  // let posts = await Post.find({ paidOut: false, payoutTime: { $lt: now } });
  let updatedPosts = posts.map(async post => {
    const average = community.currentShares / community.postCount;

    // only reward above-average posts
    console.log('rank vs average ', post.pagerank, average);
    if (post.pagerank < average) {
      post.payout = 0;
      return post.save();
    }
    // linear reward curve
    post.payoutShare = post.pagerank / (community.topPostShares || 1);
    post.payout = community.rewardFund * post.payoutShare;
    return post.save();
  });
  updatedPosts = await Promise.all(updatedPosts);
  return updatedPosts;
}

async function allocateRewards() {
  await Eth.mintRewardTokens();
  const rewardPool = await Eth.getParam('rewardPool', { noConvert: true });
  return rewardPool;
}

async function distributeRewards(community, rewardPool) {
  // let treasury = await Treasury.findOne({ community });
  const now = new Date();

  // use postData as post
  const posts = await PostData.find({
    eligibleForReward: true,
    paidOut: false,
    payoutTime: { $lte: now },
    communityId: community._id
  });

  const pendingPayouts = await PostData.find({
    eligibleForReward: true,
    paidOut: false,
    payoutTime: { $gt: now },
    communityId: community._id
  });

  // decay current reward shares
  const decay = (now.getTime() - community.lastRewardFundUpdate.getTime()) / SHARE_DECAY;

  community.rewardFund = rewardPool;
  community.currentShares *= 1 - Math.min(1, decay);
  community.postCount *= 1 - Math.min(1, decay);
  console.log('total shares ', community.currentShares);

  // add post relevance to treasury
  posts.forEach(post => {
    // cut off low-ranking posts
    if (post.pagerank > MINIMUM_RANK) {
      community.currentShares += post.pagerank;
      community.postCount += 1;
      if (post.pagerank >= community.currentShares / community.postCount) {
        community.topPostShares += post.pagerank;
      }
      console.log('average post shares ', community.currentShares / community.postCount);
    }
    post.paidOut = true;
  });

  community.lastRewardFundUpdate = now;
  community = await community.save();

  const updatedPosts = await computePostPayout(posts, community);
  // estimates post payout
  await computePostPayout(pendingPayouts, community);

  return updatedPosts;
}

async function rewardUser(props) {
  const { user, reward, post, community, communityId, type } = props;

  // do we need checks against reward fund? don't think so since it comes from smart contract
  // treasury.rewardFund -= reward;
  // if (treasury.rewardFund < 0) throw new Error('Reward fund is empty!');
  // treasury = await treasury.save();
  // user.balance += reward;
  // user = await user.save();

  console.log('Awarded ', user.name, ' ', reward, ' tokens for ', post);

  const s = reward === 1 ? '' : 's';
  const action = type === 'vote' ? 'upvoting ' : '';

  const amount = numbers.abbreviateNumber(reward);
  // let amount = reward;
  const text = `You earned ${amount} coin${s} from ${action}this post`;
  const alertText = `You earned ${amount} coin${s} from ${action}a post`;
  console.log(text);

  await Earnings.updateRewardsRecord({
    user: user._id,
    post,
    earned: reward,
    status: 'paidout',
    community,
    communityId,
  });

  Notification.createNotification({
    post,
    forUser: user._id,
    type: 'reward',
    coin: reward,
    text,
    coinType: 'eth',
    community,
    communityId
  });

  apnData.sendNotification(user, alertText, {});
  return user;
}

async function distributeUserRewards(posts, community) {
  const { slug, communityId } = community;
  const payouts = {};
  const notifications = [];
  let distributedRewards = 0;

  const updatedUsers = posts.map(async post => {
    const votes = await Invest.find({ post: post.post });
    // compute total vote shares

    let totalShares = 0;
    votes.forEach(vote => {
      totalShares += vote.shares;
    });

    if (totalShares === 0) {
      return null;
    }

    console.log('totalShares ', totalShares);
    console.log('post shares ', post.shares);
    const curationReward = post.payout;

    //  ---------- Curation rewards ------------

    console.log('curationReward', curationReward);

    const updatedVotes = votes.map(async vote => {
      // don't count downvotes
      if (vote.amount < 0) return;

      const user = await User.findOne(
        { _id: vote.investor },
        'name balance deviceTokens badge'
      );

      const curationWeight = vote.shares / totalShares;
      const curationPayout = Math.floor(curationWeight * curationReward);

      distributedRewards += curationPayout;

      console.log('weight ', curationWeight);
      console.log('payout ', curationPayout);
      if (curationPayout === 0) return;

      payouts[user._id] = payouts[user._id]
        ? payouts[user._id] + curationPayout
        : curationPayout;

      // TODO diff decimal
      user.balance += curationPayout / 10 ** 18;
      await user.save();

      notifications.push({
        user,
        reward: curationPayout / 10 ** 18,
        // treasury,
        post: post.post,
        type: 'vote',
        community: slug,
        communityId,
      });
    });
    return Promise.all(updatedVotes);
  });

  await Promise.all(updatedUsers);

  // transfer amounts to distributed rewards
  console.log('distribute rewards ', distributedRewards);
  console.log('distribute rewards should be', distributedRewards);

  // we'll do this individually upon request to save on gas
  const sendNotes = notifications.map(async n => rewardUser(n));
  await Promise.all(sendNotes);

  totalDistributedRewards += distributedRewards;

  return payouts;
}

// exports.pendingUserReward(userId) {
//   let now = new Date();
//   let votes = await Invest.find({ payoutDate: { $lt: now } });
// }
//

async function computeCommunityRewards(community, _rewardPool, balances) {
  await computePageRank({ communityId: community._id, community: community.slug });

  const totalBalance = balances.reduce((a, c) => c.balance + a, 0);
  const communityBalance = balances.find(c => c._id === community.slug).balance;

  // compute portion of reward pool allocated to community
  const communityRewardPool = (_rewardPool * communityBalance) / totalBalance;

  console.log('Rewards for ', community.slug, ' ', communityRewardPool);

  const updatedPosts = await distributeRewards(community, communityRewardPool);
  const payouts = await distributeUserRewards(updatedPosts, community);

  const displayPosts = updatedPosts.map(p => ({
    title: p.post,
    payout: p.payout,
    payoutShare: p.payoutShare,
    rank: p.pagerank
  }));

  console.log(
    '\x1b[32m',
    'allocated rewards to ',
    community.slug,
    ' ',
    communityRewardPool
  );
  console.log('\x1b[32m', 'distributed rewards to posts ', displayPosts);
  console.log('\x1b[32m', payouts);
  console.log('\x1b[0m');
  return payouts;
}

let computingRewards = false;

exports.rewards = async () => {
  // safeguard
  if (computingRewards) {
    throw new Error('computing rewards is already in progress!');
  }
  computingRewards = true;
  try {
    totalDistributedRewards = 0;
    let communities = await Community.find({});
    let rewardPool = await allocateRewards();
    const balances = await Community.getBalances();
    communities.forEach(c =>
      q.push(async cb => {
        try {
          const res = await computeCommunityRewards(c, rewardPool, balances);
          return cb(null, res);
        } catch (err) {
          throw new Error(err);
        }
      })
    );

    // start loop
    communities = await new Promise((resolve, reject) =>
      q.start((err, results) => {
        if (err) reject(err);
        resolve(results);
      })
    );
    computingRewards = false;

    if (totalDistributedRewards > 0) {
      await Eth.allocateRewards(totalDistributedRewards);
    }

    rewardPool = await Eth.getParam('rewardPool', { noConvert: true });
    await Eth.getParam('distributedRewards', { noConvert: true });

    // console.log('distributedRewards Pool', distPool);
    console.log('Finished distributing rewards, remaining reward fund: ', rewardPool);

    return communities;
  } catch (error) {
    console.log('rewards error', error);
    computingRewards = true;
    return null;
  }
};
