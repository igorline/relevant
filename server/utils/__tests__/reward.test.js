import ethRewards from 'server/utils/ethRewards';
import { sanitize } from 'server/test/utils';
import Invest from 'server/api/invest/invest.model';
import User from 'server/api/user/user.model';
import Post from 'server/api/post/post.model';
import PostData from 'server/api/post/postData.model';
import Earnings from 'server/api/earnings/earnings.model';
import Notification from 'server/api/notification/notification.model';
import { getUsers, getPosts, getCommunities } from 'server/test/seedData';
import * as Eth from 'server/utils/ethereum';
import Treasury from 'server/api/treasury/treasury.model';

// this will define the database name where the tests are run
process.env.TEST_SUITE = 'eth-rewards';
let { alice, bob, carol, relevant, crypto, link1, link2, link3, link4, link5 } = {};

jest.mock('server/utils/ethereum');

describe('ethRewards', () => {
  Eth.mintRewardTokens = jest.fn();
  Eth.allocateRewards = jest.fn(async () => Treasury.findOne({ community: 'global' }));
  Eth.getParam.mockImplementation(() => 10000 * 1e18);

  beforeEach(() => {
    ({ alice, bob, carol } = getUsers());
    ({ link1, link2, link3, link4, link5 } = getPosts());
    ({ relevant, crypto } = getCommunities());
    global.console = { log: jest.fn() }; // hides logs
  });

  describe('No Rewards', () => {
    test('should not compute rewards', async () => {
      const payouts = await ethRewards.rewards();
      expect(payouts).toBe(false);
    });
  });

  describe('Votes', () => {
    test('should be created', async () => {
      const voteArray = createVoteArray();

      const wait = {};
      wait.p = Promise.resolve(true);
      const votes = await voteArray.reduce(async (executed, vote) => {
        executed = await executed;
        return [...executed, await voteAndBet(vote)];
      }, Promise.resolve([]));

      const bet = sanitize(votes[0].bet);
      expect(bet).toMatchSnapshot();
    });
  });

  describe('Rewards', () => {
    test('should compute correctly', async () => {
      await crypto.join(carol._id);
      const payouts = await ethRewards.rewards();
      Object.values(payouts.payoutData).forEach(v => {
        v.distributedRewards = v.distributedRewards.toPrecision(12);
      });
      payouts.totalDistributedRewards = payouts.totalDistributedRewards.toPrecision(12);
      expect(payouts).toMatchSnapshot();
      expect(Eth.allocateRewards).toHaveBeenCalled();
      expect(Eth.mintRewardTokens).toHaveBeenCalled();
      const treasury = await Eth.allocateRewards.mock.results[0].value;
      expect(treasury.unAllocatedRewards.toPrecision(12)).toBe(
        payouts.totalDistributedRewards
      );
      const afterAllocation = await Treasury.findOne({ community: 'global' });
      expect(afterAllocation.unAllocatedRewards).toBe(0);
    });
  });

  describe('After rewards', () => {
    test('earnings should update', async () => {
      let earnings = await Earnings.findOne({ post: link1._id, user: alice._id });
      expect(earnings.status).toBe('paidout');
      expect(earnings.community).toBe('relevant');

      earnings = sanitize(earnings);
      expect(earnings).toMatchSnapshot();

      const earningBob = await Earnings.findOne({ post: link1._id, user: bob._id });
      expect(earningBob.status).toBe('paidout');
      expect(earningBob.community).toBe('crypto');
    });

    test('cross post payout should compute correctly', async () => {
      const postDataR = await PostData.findOne({
        post: link1._id,
        communityId: relevant._id
      });
      const postDataC = await PostData.findOne({
        post: link1._id,
        communityId: crypto._id
      });
      const earningAlice = await Earnings.findOne({ post: link1._id, user: alice._id });
      const earningBob = await Earnings.findOne({ post: link1._id, user: bob._id });

      const rewardAlice =
        (postDataR.payout * earningAlice.shares * 1e-18) / postDataR.shares;
      const rewardBob = (postDataC.payout * earningBob.shares * 1e-18) / postDataC.shares;

      expect(rewardAlice).toBeCloseTo(earningAlice.earned, 10);
      expect(rewardBob).toBeCloseTo(earningBob.earned, 10);
    });

    test('invest shares and post shares should match', async () => {
      const postDataR = await PostData.findOne({
        post: link1._id,
        communityId: relevant._id
      });
      const postDataC = await PostData.findOne({
        post: link1._id,
        communityId: crypto._id
      });

      const investR = await Invest.find({ post: link1._id, communityId: relevant._id });
      const investC = await Invest.find({ post: link1._id, communityId: crypto._id });

      const sharesR = investR.reduce((a, v) => a + v.shares, 0);
      const sharesC = investC.reduce((a, v) => a + v.shares, 0);

      expect(sharesR).toBe(postDataR.shares);
      expect(sharesC).toBe(postDataC.shares);
    });

    test('locked tokens should be returned', async () => {
      const user = await User.findOne({ _id: alice._id }, 'lockedTokens');
      expect(user.lockedTokens).toBe(0);
    });

    test('nopayout locked tokens should be returned', async () => {
      const user = await User.findOne({ _id: carol._id }, 'lockedTokens');
      expect(user.lockedTokens).toBe(0);
    });

    test('nopayout earning should be expired', async () => {
      const earning = await Earnings.findOne(
        { user: carol._id, post: link5._id },
        'status'
      );
      expect(earning.status).toBe('expired');
    });

    test('mature posts should be paidout', async () => {
      const post = await Post.findOne({ _id: link1._id }).populate('data');
      expect(post.data.paidOut).toBe(true);
    });

    test('pending posts should not be paid out', async () => {
      const post = await Post.findOne({ _id: link4._id }).populate('data');
      expect(post.data.paidOut).toBe(false);
    });

    test('pending posts should update', async () => {
      let pending = await Earnings.findOne({ post: link4._id, user: bob._id });
      expect(pending.estimatedPostPayout > 0).toBeTruthy();
      pending = sanitize(pending);
      expect(pending).toMatchSnapshot();
    });

    test('should create notification', async () => {
      let notification = await Notification.findOne({
        post: link1._id,
        forUser: alice._id
      });
      notification = sanitize(notification);
      expect(notification).toMatchSnapshot();
    });

    test('should reset community staked tokens', async () => {
      const postStake = await PostData.findOne({ post: link4._id });
      const updatedBalances = await Earnings.stakedTokens();
      const totalBalance = updatedBalances.reduce((a, c) => c.stakedTokens + a, 0);
      expect(totalBalance).toBe(postStake.balance);
    });

    // test('logs should match', async () => {
    //   const logs = console.log.mock.calls
    //   .map(log => log.map(l => JSON.stringify(l)).join(' ').trim())
    //   .join('\n').replace(/"/g, '');
    //   expect(logs).toMatchSnapshot();
    // });
  });

  async function voteAndBet(vote) {
    const repVote = await Invest.createVote(vote);
    const stakedTokens = (vote.user.balance + vote.user.tokenBalance) * 0.1;
    const bet = await repVote.placeBet({ ...vote, stakedTokens });
    return { repVote, bet };
  }

  function createVoteArray() {
    const relevantVote = {
      communityInstance: relevant,
      community: relevant.slug,
      communityId: relevant._id,
      amount: 1
    };

    const cryptoVote = {
      communityInstance: crypto,
      community: crypto.slug,
      communityId: crypto._id,
      amount: 1
    };

    return [
      { ...relevantVote, user: alice, post: link1 },
      { ...cryptoVote, user: bob, post: link1 },

      { ...relevantVote, user: bob, post: link2 },
      { ...relevantVote, user: carol, post: link2 },
      { ...cryptoVote, user: bob, post: link3 },
      { ...cryptoVote, user: bob, post: link4 },
      { ...cryptoVote, user: carol, post: link5 }
    ];
  }
});

// async function placeBet({ user, post }) {
//   const stakedTokens = (user.balance + user.tokenBalance ) * 0.1;
//   await
// }
