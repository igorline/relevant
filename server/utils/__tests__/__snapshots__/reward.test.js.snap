// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ethRewards After rewards earnings should update 1`] = `
Object {
  "__v": 0,
  "_id": "[SANITIZED]",
  "airdropTokens": 111,
  "cashOutAmt": 0,
  "cashOutAttempt": false,
  "community": "relevant",
  "communityId": "c00000000000000000000001",
  "createdAt": "[SANITIZED]",
  "earned": 158.82020314845198,
  "endBalance": 269.820203148452,
  "estimatedPostPayout": 0,
  "legacyAirdrop": 0,
  "payoutTime": "[SANITIZED]",
  "post": "000000000000000000000003",
  "prevBalance": 111,
  "referralTokens": 0,
  "shares": 4.711687595755899,
  "source": "post",
  "spent": 0,
  "stakedTokens": 11.100000000000001,
  "status": "paidout",
  "totalPostShares": 4.711687595755899,
  "totalPreviousPaidout": 0,
  "type": "coins",
  "updatedAt": "[SANITIZED]",
  "user": "a00000000000000000000000",
}
`;

exports[`ethRewards After rewards pending posts should update 1`] = `
Object {
  "__v": 0,
  "_id": "[SANITIZED]",
  "airdropTokens": 0,
  "cashOutAmt": 0,
  "cashOutAttempt": false,
  "community": "crypto",
  "communityId": "c00000000000000000000002",
  "createdAt": "[SANITIZED]",
  "earned": 0,
  "endBalance": 0,
  "estimatedPostPayout": 271.8,
  "legacyAirdrop": 0,
  "payoutTime": "[SANITIZED]",
  "post": "000000000000000000000006",
  "prevBalance": 0,
  "referralTokens": 0,
  "shares": 6.663332499583073,
  "source": "post",
  "spent": 0,
  "stakedTokens": 22.200000000000003,
  "status": "pending",
  "totalPostShares": 6.663332499583073,
  "totalPreviousPaidout": 0,
  "type": "coins",
  "updatedAt": "[SANITIZED]",
  "user": "b00000000000000000000000",
}
`;

exports[`ethRewards After rewards should create notification 1`] = `
Object {
  "__v": 0,
  "_id": "[SANITIZED]",
  "byUsers": Array [],
  "byUsersHandle": Array [],
  "coin": 158.82020314845198,
  "coinType": "eth",
  "createdAt": "[SANITIZED]",
  "forUser": "a00000000000000000000000",
  "group": Array [],
  "personal": true,
  "post": "000000000000000000000003",
  "read": false,
  "source": "post",
  "text": "You earned 159 coins from upvoting this post",
  "type": "reward",
  "updatedAt": "[SANITIZED]",
}
`;

exports[`ethRewards Rewards should compute correctly 1`] = `
Object {
  "payoutData": Object {
    "crypto": Object {
      "community": "crypto",
      "distributedRewards": "6.04000000000e+20",
      "payouts": Object {
        "b00000000000000000000000": 604000000000000100000,
        "c00000000000000000000000": 0,
      },
      "postPayouts": Array [
        Object {
          "payout": 302000000000000070000,
          "payoutShare": 0.05,
          "rank": 100,
          "title": "000000000000000000000003",
        },
        Object {
          "payout": 302000000000000070000,
          "payoutShare": 0.05,
          "rank": 100,
          "title": "000000000000000000000005",
        },
        Object {
          "payout": 0,
          "payoutShare": undefined,
          "rank": 0,
          "title": "000000000000000000000007",
        },
      ],
    },
    "relevant": Object {
      "community": "relevant",
      "distributedRewards": "3.56820203148e+20",
      "payouts": Object {
        "a00000000000000000000000": 158820203148451970000,
        "b00000000000000000000000": 125226195342667870000,
        "c00000000000000000000000": 72773804657332210000,
      },
      "postPayouts": Array [
        Object {
          "payout": 158820203148451970000,
          "payoutShare": 0.04010611190617473,
          "rank": 45.96,
          "title": "000000000000000000000003",
        },
        Object {
          "payout": 198000000000000070000,
          "payoutShare": 0.05,
          "rank": 100,
          "title": "000000000000000000000004",
        },
      ],
    },
  },
  "totalDistributedRewards": "9.60820203148e+20",
}
`;

exports[`ethRewards Votes should be created 1`] = `
Object {
  "__v": 0,
  "_id": "[SANITIZED]",
  "amount": 1,
  "community": "relevant",
  "communityId": "c00000000000000000000001",
  "createdAt": "[SANITIZED]",
  "downVotes": 0,
  "investor": "a00000000000000000000000",
  "isManualBet": true,
  "ownPost": false,
  "paidOut": false,
  "partialRelevance": 0,
  "partialUsers": 0,
  "post": "000000000000000000000003",
  "rankChange": 0,
  "relevance": 0,
  "relevantPoints": 0,
  "shares": 4.711687595755899,
  "stakedTokens": 11.100000000000001,
  "updatedAt": "[SANITIZED]",
  "upvotes": 0,
}
`;
